src/puppet.ts

```diff
@@ -378,8 +362,71 @@ export abstract class Puppet extends PuppetEventEmitter {
    *
    *
    */
-  public abstract async start () : Promise<void>
-  public abstract async stop ()  : Promise<void>
+  public async start () : Promise<void> {
+    this.on('heartbeat', this.feedDog)
+    this.watchdog.on('reset', this.dogReset)
+
+    /**
+     * 2. Setup `reset` Event via a 1 second Throttle Queue:
+     */
+    this.resetThrottleQueue = new ThrottleQueue<string>(1000)
+    // 2.2. handle all `reset` events via the resetThrottleQueue
+    this.on('reset', this.throttleReset)
+    // 2.3. call reset() and then ignore the following `reset` event for 1 second
+    this.resetThrottleQueueSubscription = this.resetThrottleQueue.subscribe(reason => {
+      log.silly('Puppet', 'constructor() resetThrottleQueue.subscribe() reason: "%s"', reason)
+      this.reset(reason)
+    })
+  }
+
+  public async stop (): Promise<void> {
+    this.removeListener('heartbeat', this.feedDog)
+    this.watchdog.removeListener('reset', this.dogReset)
+    this.removeListener('reset', this.throttleReset)
+
+    if (this.resetThrottleQueueSubscription) {
```

Comment:
If we keep the subscription between `start()` and `stop()`, then this part of the code can be kept inside the `constructor()` too.

Reply:
Currently we are always calling the `start()` after the `stop()`, but it might also have cases that the user only call the stop, to achieve that, I extract the queue and subscriptions into `start()` and `stop()`.

Reply:
Could you explain why a user should care about whether this subscription exists after the puppet is stopped?

I think it should be ok that we can encapsulate this subscription inside the puppet and keep it subscripted after the puppet instantiated.

Reply:
> Could you explain why a user should care about whether this subscription exists after the puppet is stopped?
I don't think user should care about the subscription anytime, these logic should be transparent to users.

> I think it should be ok that we can encapsulate this subscription inside the puppet and keep it subscripted after the puppet instantiated.

We can do this, it's just feels like not everything is cleared after the `stop()` call.

But I won't insist on this issue, I just expressed the reason behind my design. Will move this to `constructor()`

Reply:
An RxJS subscript is very clean, so let's just set-and-forget it.